<a href="https://colab.research.google.com/github/VigneshJ19/ML-LAB/blob/main/Simple%20ML%20Model%20ex%201.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **	Implementation of Simple ML Model**

NAME : Vignesh J 

REG.NO : 20MAI0012

In [ ]:
import sklearn.datasets
import numpy as np

In [ ]:
cancer_ds = sklearn.datasets.load_breast_cancer()

In [ ]:
X = cancer_ds.data 
Y = cancer_ds.target

In [ ]:
print(X)
print(Y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [ ]:
print(X.shape, Y.shape)

(569, 30) (569,)


In [ ]:
import pandas as pd

In [ ]:
data = pd.DataFrame(cancer_ds.data, columns = cancer_ds.feature_names)

In [ ]:
data['class'] = cancer_ds.target

In [ ]:
data.head()

mean radius  mean texture  ...  worst fractal dimension  class
0        17.99         10.38  ...                  0.11890      0
1        20.57         17.77  ...                  0.08902      0
2        19.69         21.25  ...                  0.08758      0
3        11.42         20.38  ...                  0.17300      0
4        20.29         14.34  ...                  0.07678      0

[5 rows x 31 columns]

In [ ]:
data.describe()

mean radius  mean texture  ...  worst fractal dimension       class
count   569.000000    569.000000  ...               569.000000  569.000000
mean     14.127292     19.289649  ...                 0.083946    0.627417
std       3.524049      4.301036  ...                 0.018061    0.483918
min       6.981000      9.710000  ...                 0.055040    0.000000
25%      11.700000     16.170000  ...                 0.071460    0.000000
50%      13.370000     18.840000  ...                 0.080040    1.000000
75%      15.780000     21.800000  ...                 0.092080    1.000000
max      28.110000     39.280000  ...                 0.207500    1.000000

[8 rows x 31 columns]

In [ ]:
print(data['class'].value_counts() )#binary variable

1    357
0    212
Name: class, dtype: int64


In [ ]:
# target names
print(cancer_ds.target_names)

['malignant' 'benign']


In [ ]:
data.groupby('class').mean()

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
class                             ...                                         
0        17.462830     21.604906  ...        0.323468                 0.091530
1        12.146524     17.914762  ...        0.270246                 0.079442

[2 rows x 30 columns]

In [ ]:
data.groupby('mean symmetry').mean()

mean radius  mean texture  ...  worst fractal dimension  class
mean symmetry                             ...                                
0.1060               11.20         29.37  ...                  0.05905    1.0
0.1167               11.41         14.92  ...                  0.07427    1.0
0.1203               12.05         22.72  ...                  0.09349    1.0
0.1215               12.49         16.85  ...                  0.06174    1.0
0.1220               17.85         13.23  ...                  0.05871    1.0
...                    ...           ...  ...                      ...    ...
0.2655               24.25         20.20  ...                  0.08009    0.0
0.2678               11.80         16.58  ...                  0.10300    0.0
0.2743               10.17         14.88  ...                  0.08020    1.0
0.2906               20.18         23.97  ...                  0.09964    0.0
0.3040               17.14         16.40  ...                  0.10590    0.0

[432 rows x 30 columns]

#Train test split

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X = data.drop('class', axis=1)
Y = data['class']

In [ ]:
# to split
X_train, X_test,Y_train, Y_test = train_test_split(X,Y)

In [ ]:
print(Y_train.shape, Y_test.shape, Y.shape)

(426,) (143,) (569,)


In [ ]:
X_train, X_test,Y_train, Y_test = train_test_split(X,Y, test_size=0.1, stratify =  Y)

In [ ]:
print(Y.mean(), Y_train.mean(), Y_test.mean())

0.6274165202108963 0.626953125 0.631578947368421


In [ ]:
print(X.mean(), X_train.mean(), X_test.mean())

mean radius                 14.127292
mean texture                19.289649
mean perimeter              91.969033
mean area                  654.889104
mean smoothness              0.096360
mean compactness             0.104341
mean concavity               0.088799
mean concave points          0.048919
mean symmetry                0.181162
mean fractal dimension       0.062798
radius error                 0.405172
texture error                1.216853
perimeter error              2.866059
area error                  40.337079
smoothness error             0.007041
compactness error            0.025478
concavity error              0.031894
concave points error         0.011796
symmetry error               0.020542
fractal dimension error      0.003795
worst radius                16.269190
worst texture               25.677223
worst perimeter            107.261213
worst area                 880.583128
worst smoothness             0.132369
worst compactness            0.254265
worst concav

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3,random_state = 0) # zero or any other integer
print(Y.mean(),Y_train.mean(),Y_test.mean())

0.6274165202108963 0.6256281407035176 0.631578947368421


In [ ]:
X_binarised_train = X_train.apply(pd.cut, bins=2, labels=[1,0])
X_binarised_test = X_test.apply(pd.cut, bins=2, labels=[1,0])
type(X_binarised_train)

pandas.core.frame.DataFrame

In [ ]:
X_binarised_train = X_binarised_test.values
X_binarised_test = X_binarised_test.values
type(X_binarised_train)

numpy.ndarray

In [ ]:
b=10

from random import randint
i = randint(0, X_binarised_train.shape[0])

if(np.sum(X_binarised_train[i,:])>=b):
    print('Model prediction is malignant')
else:
    print('Model prediction is benign')

if(Y_train[i]==1):
     print('Actual prediction is malignant')
else:
    print('Actual prediction is benign')


Model prediction is malignant
Actual prediction is malignant


In [ ]:
for b in range (X_binarised_train.shape[1]+1):
    accurate_rows = 0
    for x,y in zip(X_binarised_train,Y_train):
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
    print("b =", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])


b = 0 111 0.6491228070175439
b = 1 111 0.6491228070175439
b = 2 111 0.6491228070175439
b = 3 111 0.6491228070175439
b = 4 111 0.6491228070175439
b = 5 111 0.6491228070175439
b = 6 111 0.6491228070175439
b = 7 111 0.6491228070175439
b = 8 112 0.6549707602339181
b = 9 112 0.6549707602339181
b = 10 113 0.6608187134502924
b = 11 112 0.6549707602339181
b = 12 111 0.6491228070175439
b = 13 111 0.6491228070175439
b = 14 111 0.6491228070175439
b = 15 110 0.6432748538011696
b = 16 109 0.6374269005847953
b = 17 106 0.6198830409356725
b = 18 103 0.6023391812865497
b = 19 102 0.5964912280701754
b = 20 101 0.5906432748538012
b = 21 99 0.5789473684210527
b = 22 101 0.5906432748538012
b = 23 98 0.5730994152046783
b = 24 98 0.5730994152046783
b = 25 98 0.5730994152046783
b = 26 96 0.5614035087719298
b = 27 92 0.5380116959064327
b = 28 85 0.49707602339181284
b = 29 83 0.4853801169590643
b = 30 75 0.43859649122807015


In [ ]:
b=27
accurate_rows = 0
    
for x,y in zip(X_binarised_train,Y_train):
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)

print("b =", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b=28
accurate_rows = 0
    
for x,y in zip(X_binarised_train,Y_train):
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)

print("b =", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b = 27 92 0.5380116959064327
b = 28 85 0.49707602339181284


In [ ]:
b=24
accurate_rows = 0
    
for x,y in zip(X_binarised_train,Y_train):
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)

print("b =", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b = 24 98 0.5730994152046783


In [ ]:
for b in range (X_binarised_test.shape[1]+1):
    accurate_rows = 0
    for x,y in zip(X_binarised_test,Y_test):
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
    print( b, accurate_rows, accurate_rows/X_binarised_test.shape[0])

0 108 0.631578947368421
1 108 0.631578947368421
2 108 0.631578947368421
3 108 0.631578947368421
4 108 0.631578947368421
5 108 0.631578947368421
6 108 0.631578947368421
7 108 0.631578947368421
8 109 0.6374269005847953
9 109 0.6374269005847953
10 110 0.6432748538011696
11 111 0.6491228070175439
12 112 0.6549707602339181
13 112 0.6549707602339181
14 112 0.6549707602339181
15 113 0.6608187134502924
16 114 0.6666666666666666
17 121 0.7076023391812866
18 126 0.7368421052631579
19 127 0.7426900584795322
20 128 0.7485380116959064
21 132 0.7719298245614035
22 136 0.7953216374269005
23 141 0.8245614035087719
24 147 0.8596491228070176
25 149 0.8713450292397661
26 151 0.8830409356725146
27 155 0.9064327485380117
28 148 0.8654970760233918
29 136 0.7953216374269005
30 108 0.631578947368421


CONCLUSION :


When training set b = 24 , Got accuracy = 0.573 and accurate row = 98.

Testing at accurate rows = 155 , Got accuracy = 0.906. 